src/theory/quantifiers/sygus/cegis_core_connective.cpp

```diff
@@ -0,0 +1,843 @@
+                                   std::vector<Node>& vals) const
+{
+  for (const Node& v : d_vars)
+  {
+    Node mv = Node::fromExpr(smt.getValue(v.toExpr()));
+    Trace("sygus-ccore-model") << v << " -> " << mv << " ";
+    vals.push_back(mv);
+  }
+}
+
+bool CegisCoreConnective::getUnsatCore(SmtEngine& smt,
+                                       Node query,
+                                       std::vector<Node>& uasserts) const
+{
+  UnsatCore uc = smt.getUnsatCore();
+  bool hasQuery = false;
+  for (UnsatCore::const_iterator i = uc.begin(); i != uc.end(); ++i)
+  {
+    Node uassert = Node::fromExpr(*i);
+    Trace("sygus-ccore-debug") << "  uc " << uassert << std::endl;
+    if (uassert == query)
+    {
+      hasQuery = true;
+      continue;
+    }
+    uasserts.push_back(uassert);
+  }
+  return hasQuery;
+}
+
+Result CegisCoreConnective::checkSat(Node n, std::vector<Node>& mvs) const
+{
+  Assert(mvs.empty());
+  Assert(n.getType().isBoolean());
+  Trace("sygus-ccore-debug") << "...check-sat " << n << "..." << std::endl;
+  n = Rewriter::rewrite(n);
+  if (n.isConst())
+  {
+    if (n.getConst<bool>())
+    {
+      // default model
+      for (const Node& v : d_vars)
+      {
+        mvs.push_back(v.getType().mkGroundTerm());
+      }
+      return Result(Result::SAT);
+    }
+    else
+    {
+      return Result(Result::UNSAT);
+    }
+  }
+  SmtEngine smt(NodeManager::currentNM()->toExprManager());
+  smt.setIsInternalSubsolver();
+  smt.setLogic(smt::currentSmtEngine()->getLogicInfo());
+  smt.assertFormula(n.toExpr());
+  Result r = smt.checkSat();
+  Trace("sygus-ccore-debug") << "...got " << r << std::endl;
+  if (r.asSatisfiabilityResult().isSat() == Result::SAT)
+  {
+    getModel(smt, mvs);
+  }
+  return r;
+}
+
+Node CegisCoreConnective::evaluate(Node n,
+                                   Node id,
+                                   const std::vector<Node>& mvs)
+{
+  Kind nk = n.getKind();
+  if (nk == AND || nk == OR)
+  {
+    NodeManager* nm = NodeManager::currentNM();
+    bool expRes = nk == OR;
+    // split AND/OR
+    for (const Node& nc : n)
+    {
+      Node enc = evaluate(nc, id, mvs);
+      Assert(enc.isConst());
+      if (enc.getConst<bool>() == expRes)
+      {
+        return nm->mkConst(expRes);
+      }
+    }
+    return nm->mkConst(!expRes);
+  }
+  std::unordered_map<Node, Node, NodeHashFunction>& ec = d_eval_cache[n];
+  if (!id.isNull())
+  {
+    std::unordered_map<Node, Node, NodeHashFunction>::iterator it = ec.find(id);
+    if (it != ec.end())
+    {
+      return it->second;
+    }
+  }
+  // use evaluator
+  Node cn = d_eval.eval(n, d_vars, mvs);
+  if (cn.isNull())
+  {
+    Node cn =
+        n.substitute(d_vars.begin(), d_vars.end(), mvs.begin(), mvs.end());
+    cn = Rewriter::rewrite(cn);
+  }
+  if (!id.isNull())
+  {
+    ec[id] = cn;
+  }
+  return cn;
+}
+
+Node CegisCoreConnective::constructSolutionFromPool(Component& ccheck,
```

Comment:
I am having a bit trouble relating this function to the algorithm that is so nicely described in the h file. It would be really useful if you include some pointers to that algorithm as comments. In particular, where is the set D constructed (both in the interpolation and in the abduction variants)?

Reply:
Ive added some pointers, let me know if its clear now.